### PR TITLE
Remove URL update functionality

### PR DESCRIPTION
**Pull request by sdlime**
This backs out URL update functionality. It's just not not as relevant as is may have been at one time. It also removes the DATAPATTERN and TEMPLATEPATTERN keywords since they are not required.

Note this initial implementation doesn't attempt to fill any gaps so 30 regression tests are failing at the moment. I'll post a subsequent comment that summarizes the failures and proposes a few targeted fixes.

See also the discussion at https://github.com/MapServer/MapServer/pull/6426

### CODE REVIEW

**Comment by sdlime** [comment]
Looking at the broken tests, the vast majority are for map.imagetype and map.resolution. So I would propose adding:

- resolution/map.resolution/map_resolution as a vendor parameter, so it applies to CGI and WXS calls
- imagetype/map.imagetype/map_imagetype as a CGI parameter only (WXS already as "format")

and immediately deprecating the map. and map_ versions so things can be cleaned up in subsequent versions. 

There was one set of tests that also set legend KEYSIZE. Keeping that syntax is not reasonable but I can see at least a little utility in tweaking that setting so I would propose adding a vendor parameter called iconsize - playing off _mode=legendicon&icon=[layer],[class]_ syntax and updating the test accordingly.

There was one test that changes the legend FONT. I propose removing that test... If there is value in that functionality we could extend runsubs to the legend object and add a validation block but I don't think it's worth doing now.

Other gaps may need to be filled but these were the obvious ones. Thoughts?

--Steve

**Review by geographika — APPROVED** [review]

**Comment by geographika** [comment]
@sdlime - the pull request is nearly all code removal, so looks good to me. I've never really used the functionality, so your suggestions for removing all updates via URLs and moving the others to named querystring parameters all sound fine to me. 

I'm not sure why I didn't remove `DATAPATTERN` and `TEMPLATEPATTERN` as part of [RFC 133](https://mapserver.org/development/rfc/ms-rfc-133.html#map). I can add them there for reference though. 

`DATA` can be replaced by [run-time substitution](https://mapserver.org/cgi/runsub.html#parameters-supported), but I don't see `TEMPLATE` listed there? Can it be changed by a parameter?

**Comment by sdlime** [comment]
> @sdlime - the pull request is nearly all code removal, so looks good to me. I've never really used the functionality, so your suggestions for removing all updates via URLs and moving the others to named querystring parameters all sound fine to me.
> 
> I'm not sure why I didn't remove `DATAPATTERN` and `TEMPLATEPATTERN` as part of [RFC 133](https://mapserver.org/development/rfc/ms-rfc-133.html#map). I can add them there for reference though.
> 
> `DATA` can be replaced by [run-time substitution](https://mapserver.org/cgi/runsub.html#parameters-supported), but I don't see `TEMPLATE` listed there? Can it be changed by a parameter?

We'll likely need to expand run-time substitution a little - web->template would be one, layer->mask is another (there was a user request for that one). I was planning on adding those as a commit to this pull request.

**Comment by sdlime** [comment]
IMHO this is ready for review. Need a little additional test coverage: 
- imagetype allows an optional regex to limit or even turn off that parameter
- bad values
- layer mask runsub
Otherwise most of this was already covered in tests.

**Review by geographika — APPROVED** [review]

**Comment by geographika** [comment]
@sdlime - I read through all the changes, and look good to me. Would the following be accurate?

Any querystring parameters in the `commonLoadForm` function can be used when both in CGI mode and when using OWS services (WMS, WFS etc.). 

The following can be sent via the querystring (all values are case-insensitive):

- `KEYSIZE` - e.g. `&keysize=10 10` split by spaces (or URL-encoded spaces). Both width and height must be within the range `MS_LEGEND_KEYSIZE_MIN` and `MS_LEGEND_KEYSIZE_MAX`
- `CLASSGROUP` - this sets the classgroup for *all* layers in the map - see #4207
- `MAP.EXTENT` or `MAP_EXTENT` - e.g. `map_extent=0 0 100 100` split by spaces (or URL-encoded spaces). These may be replaced by a Vendor parameter `EXTENT` in the future. 
- `MAP.RESOLUTION` or `MAP_RESOLUTION` (within MS_RESOLUTION_MIN MS_RESOLUTION_MAX boundaries) `RESOLUTION` can't be used here as it has a conflict with a WCS keyword. 

What happens if these keywords are also used for runtime subsitution? Do they apply to both the Mapfile changes above and as replacements (e.g. what happens if %KEYSIZE% is used in a `LAYER` `DATA` clause?
Hopefully it applies to both cases?

`msCGILoadForm` allows lots of other querystring parameters to be used, but only the following modify the Mapfile:

- `RESOLUTION` (within MS_RESOLUTION_MIN MS_RESOLUTION_MAX boundaries)
- `IMAGETYPE`, `MAP.IMAGETYPE` and `MAP_IMAGETYPE`.

**Review by rouault — APPROVED** [review]

**Comment by sdlime** [comment]
> @sdlime - I read through all the changes, and look good to me. Would the following be accurate?

I think so, yes.

> What happens if these keywords are also used for runtime subsitution? Do they apply to both the Mapfile changes above and as replacements (e.g. what happens if %KEYSIZE% is used in a `LAYER` `DATA` clause? Hopefully it applies to both cases?

Yes, they apply in both cases. Note that KEYSIZE as interpreted as a parameter is subject to different validation rules (e.g. min/max key size values).  When used as part of a runsub it would be subject to a regex validation rule and treated as a single string. Runsub treats all the params the same, as strings subject to a specific regex validation rule.

File: msautotest/wxs/wms_legend.map [hunk 1]
```diff
@@ -19,7 +19,7 @@
...
 # RUN_PARMS: mode_maplegend_all.png [MAPSERV] QUERY_STRING='map=[MAPFILE]&mode=maplegend&mapext=2258982+-70747+2615354+495480&layer=layer1'  > [RESULT_DEMIME]
 # RUN_PARMS: mode_maplegend_74.png [MAPSERV] QUERY_STRING='map=[MAPFILE]&mode=maplegend&mapext=2291338+-73348+2301636+-64058&layer=layer1'  > [RESULT_DEMIME]
-# RUN_PARMS: wms_content_legend_l1_changefont.png [MAPSERV] QUERY_STRING="map=[MAPFILE]&SERVICE=WMS&VERSION=1.1.0&REQUEST=GetLegendGraphic&LAYER=layer1&FORMAT=image/png&BGCOLOR=0xFFFFFF&TRANSPARENT=FALSE&EXCEPTIONS=INIMAGE&map.legend.label=font+VeraBd" > [RESULT_DEMIME]
+# OBSOLETE: wms_content_legend_l1_changefont.png [MAPSERV] QUERY_STRING="map=[MAPFILE]&SERVICE=WMS&VERSION=1.1.0&REQUEST=GetLegendGraphic&LAYER=layer1&FORMAT=image/png&BGCOLOR=0xFFFFFF&TRANSPARENT=FALSE&EXCEPTIONS=INIMAGE&map.legend.label=font+VeraBd" > [RESULT_DEMIME]
```
[the remark this answers is not in the document]
Comment (sdlime) [reply msautotest/wxs/wms_legend.map:22]: Not worth hanging onto as a reminder? Wasn't sure...

[the remark this answers is not in the document]
Comment (rouault) [reply msautotest/wxs/wms_legend.map:22]: I don't really see the point in keeping relics of the past :-) At worse we can dig through git history

[the remark this answers is not in the document]
Comment (sdlime) [reply msautotest/wxs/wms_legend.map]: Done!

**Comment by geographika** [comment]
@sdlime - I'm updating the Itasca demo on the OSGeoLive demo. It makes use of dynamic TEMPLATEs set using URL update functionality, which is now removed. In the comments above `WEB->TEMPLATE` was mentioned as one of the parameters that would be updated to allow run-time substitution but I don't think it was added?
The TEMPLATE parameter always seems to remain unchamged even when a `"%template%"` value is set in the Mapfile and a value passed in via a URL.

**Pull request merged**